src/libos/crates/async-rt/src/sync/mutex/mod.rs

```diff
@@ -0,0 +1,122 @@
+use super::*;
+
+mod inner;
+
+use inner::MutexInner;
+use std::boxed::Box;
+
+/// An asynchronous mutex type.
+///
+/// This is simillar to `std::sync::Mutex` but with the following differences:
+/// - `lock` method is asynchronous and will not block
+/// - the `MutexGuard` can be held across `await` calls
+pub struct Mutex<T: ?Sized> {
+    inner: Box<MutexInner>,
+    data: UnsafeCell<T>,
+}
+
+/// Mutex can be used across threads as long as T is `Send`.
+unsafe impl<T> Send for Mutex<T> where T: ?Sized + Send {}
+unsafe impl<T> Sync for Mutex<T> where T: ?Sized + Send {}
+
+/// A handle to a held mutex. This can be used across `await` calls because
+/// it is `Send`.
+pub struct MutexGuard<'a, T: ?Sized + 'a> {
+    lock: &'a Mutex<T>,
+}
+
+/// Automatically `Send` marked for MutexGuard.
+unsafe impl<T: ?Sized + Sync> Sync for MutexGuard<'_, T> {}
+
+impl<T> Mutex<T> {
+    pub fn new(t: T) -> Mutex<T> {
+        Mutex {
+            inner: Box::new(MutexInner::new()),
+            data: UnsafeCell::new(t),
+        }
+    }
+}
+
+impl<T: ?Sized> Mutex<T> {
+    /// Async method to lock the mutex
+    pub async fn lock(&self) -> MutexGuard<'_, T> {
+        self.inner.lock().await.unwrap();
+        MutexGuard::new(self)
+    }
+
+    /// Try acquiring the lock without blocking
+    pub fn try_lock(&self) -> Result<MutexGuard<'_, T>> {
+        self.inner.try_lock()?;
+        Ok(MutexGuard::new(self))
+    }
+
+    /// Consume the mutex to get inner T
+    pub fn into_inner(self) -> T
+    where
+        T: Sized,
+    {
+        let (inner, data) = {
+            let Mutex {
+                ref inner,
+                ref data,
+            } = self;
+            unsafe { (ptr::read(inner), ptr::read(data)) }
+        };
+        mem::forget(self);
+        drop(inner);
+
+        data.into_inner()
+    }
+
+    /// Get a mutable reference to the inner data
+    pub fn get_mut(&mut self) -> &mut T {
+        unsafe { &mut *self.data.get() }
+    }
+}
+
+impl<T: core::fmt::Debug> fmt::Debug for Mutex<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Mutex lock")
+            .field("inner", &self.inner)
+            .field("data", unsafe { &(*self.data.get()) })
+            .finish()
+    }
+}
+
+impl<T: ?Sized + Default> Default for Mutex<T> {
+    fn default() -> Mutex<T> {
+        Mutex::new(Default::default())
+    }
+}
+
+impl<'a, T: ?Sized> MutexGuard<'a, T> {
+    pub fn new(lock: &'a Mutex<T>) -> MutexGuard<'a, T> {
```

Comment:
This method must be private. The users of `Mutex` must not be able to instantiate `MutexGuard` without calling `Mutex::lock` first.